### ISSUE TITLE
RMSE values significantly higher than reported in README when reproducing on individual timestamps

### ISSUE THREAD

**Issue by salmanmohebi** [taohan10200/CRA5#11]
Hi, thank you for the great work on CRA5!

I've been reproducing the per-variable RMSE values reported in the README table using the pretrained VAEformer (268-channel, quality=268) and the official `cra5_api` pipeline. I followed exactly the approach described in the README — downloading ERA5 NetCDF files, encoding to binary, decoding, and computing RMSE against the original ERA5 data. I also verified that:

- Variable ordering and pressure level mapping are correct (268 channels: 7 pressure vars × 37 levels + 9 surface vars, levels ordered 1000→1 hPa)
- Normalization statistics (`mean_std.json`, `mean_std_single.json`) match exactly between my pipeline and the API
- `net.update(force=True)` was called to initialize the CDFs
- The full binary encode→decode path produces identical results to the direct forward pass

However, the RMSE I observe on individual timestamps is substantially higher than the values in the README, particularly for upper-atmosphere variables. Here are some representative comparisons (using timestamp `2020-01-01T00:00:00`):

| Variable | README RMSE | My RMSE | Ratio |
|----------|-------------|---------|-------|
| z_500    | 6.098       | 16.551  | 2.71× |
| z_1      | 27.987      | 165.024 | 5.90× |
| z_20     | 16.538      | 134.536 | 8.14× |
| t_70     | 0.155       | 1.041   | 6.72× |
| t2m      | 0.720       | 1.234   | 1.71× |
| msl      | 12.685      | 22.365  | 1.76× |
| u_500    | 0.676       | 0.652   | 0.97× |
| v_500    | 0.603       | 0.590   | 0.98× |
| t_500    | 0.185       | 0.197   | 1.06× |

Overall statistics across 250 channels with README values:
- Median ratio (my RMSE / README RMSE): **1.105**
- Within 10% of README: 121/250 channels
- Within 20% of README: 151/250 channels

The pattern is clear: lower/mid-troposphere variables (500–1000 hPa) match well (ratios ~0.95–1.12), but upper-atmosphere variables (1–70 hPa), especially geopotential (z), show ratios of 5–8×.

## Questions

1. **How many timestamps were averaged over** to produce the error table in the README? Were these computed over the full test set (e.g., all of 2020, ~8760 hourly timestamps)? The README doesn't specify this, and I suspect averaging over many timestamps would reduce the apparent RMSE for high-variability upper-atmosphere fields.

2. **Is the reported error the global spatial RMSE**, or is there any latitude weighting (cosine weighting) applied? I tested latitude-weighted RMSE but it only changes results by ~1.25× at most, not enough to explain the 5–8× discrepancy.

3. **Which specific ERA5 data source was used?** I used ERA5 data from the CDS archive. Were the reference values perhaps computed against a different data version or using data downloaded through a specific pipeline?

4. **Is there any post-processing or clipping** applied to the reconstruction before computing RMSE that is not reflected in the code?

Any clarification on the evaluation protocol would be very helpful for users trying to reproduce and build upon your results. Thank you!

**Comment by Chieh997** [comment]
Hi there, thanks for raising this and sharing your evaluation setup!

We were investigating this as well and noticed that the discrepancy likely comes from the data source. When we evaluate against Google's **ARCO-ERA5** dataset instead of raw CDS downloads, our RMSE matches the original README almost perfectly. 

As fellow users, our guess is that CRA5's normalization stats (`mean_std.json`) were fitted to an ARCO-like pipeline. While both are "ERA5", subtle differences in precision, packing, and grid post-processing seem to heavily inflate CDS errors for extreme-level variables and geopotential. 

Here are the results we got evaluating against ARCO-ERA5 (`gcp-public-data-arco-era5`), using same timestamp `2020-01-01T00:00:00`:

| Variable | Unit | README RMSE | CDS RMSE (Reported) | Our ARCO RMSE | Our ARCO MAE | Rel L2 Error |
|:--------:|------|------------:|---------:|--------:|--------:|-------------:|
| z_500    | m^2/s^2    |       6.098 |   16.551 |  5.9962 |  4.1712 |     1.11e-04 |
| z_1      | m^2/s^2    |      27.987 |  165.024 | 38.2173 | 25.5130 |     8.19e-05 |
| z_20     | m^2/s^2    |      16.538 |  134.536 | 16.0143 | 11.4142 |     6.24e-05 |
| t_70     | K    |       0.155 |    1.041 |  0.1471 |  0.0987 |     6.99e-04 |
| t2m      | K    |       0.720 |    1.234 |  0.7951 |  0.4568 |     2.86e-03 |
| msl      | Pa   |      12.685 |   22.365 | 13.9652 |  7.5309 |     1.38e-04 |
| u_500    | m/s  |       0.676 |    0.652 |  0.6289 |  0.4459 |     4.44e-02 |
| v_500    | m/s  |       0.603 |    0.590 |  0.5589 |  0.3902 |     5.73e-02 |
| t_500    | K    |       0.185 |    0.197 |  0.1724 |  0.1238 |     6.82e-04 |


If you'd like to reproduce these numbers, here is the script we used to fetch the ARCO-ERA5 ground truth from GCP:


```python
import os
import torch
import xarray as xr
import numpy as np
import matplotlib.pyplot as plt
from huggingface_hub import hf_hub_download
from cra5.api import cra5_api

# Force CPU to avoid  errors
# os.environ["CUDA_VISIBLE_DEVICES"] = ""

def test_custom_9_vars(target_time="2020-01-01T00:00:00"):
    # 1. Variable mapping definition
    var_config = [
        ('z_500',  'geopotential', True, 500, 'm'),
        ('z_1',    'geopotential', True, 1,   'm'),
        ('z_20',   'geopotential', True, 20,  'm'),
        ('t_70',   'temperature',  True, 70,  'K'),
        ('t2m',    '2m_temperature', False, None, 'K'),
        ('msl',    'mean_sea_level_pressure', False, None, 'Pa'),
        ('u_500',  'u_component_of_wind', True, 500, 'm/s'),
        ('v_500',  'v_component_of_wind', True, 500, 'm/s'),
        ('t_500',  'temperature', True, 500, 'K')
    ]
    year = target_time.split('-')[0]
    temp_dir = "./temp_9vars"
    
    try:
        # 2. Download and initialize API
        print(f"📦 Downloading CRA5 data for {target_time}...")
        bin_path = hf_hub_download(
            repo_id="taohan10200/CRA5-Dataset", repo_type="dataset",
            filename=f"{year}/{target_time}.bin", local_dir=temp_dir)
        
        cra5_API = cra5_api()
        _, vname_to_channels = cra5_API.channel_vname_mapping()

        recon_output = cra5_API.decode_from_bin(time_stamp=target_time, custom_path=bin_path, return_format='de_normalized')
        big_numpy = list(recon_output.values())[0].squeeze().numpy()

        # 3. Fetch ARCO-ERA5 
        print("🌍 Connecting to GCP to fetch corresponding levels...")
        url = 'gs://gcp-public-data-arco-era5/ar/full_37-1h-0p25deg-chunk-1.zarr-v3'
        ds = xr.open_zarr(url, chunks=None, storage_options=dict(token='anon'), consolidated=True)

        # 4. Calculation
        results_table = []
        print(f"\n--- Calculating Metrics for {target_time} ---\n")

        for i, (c_name, a_name, has_level, level_val, unit) in enumerate(var_config):
            
            # --- Get Ground Truth ---
            if has_level:
                gt_data = ds[a_name].sel(time=target_time, level=level_val).values
            else:
                gt_data = ds[a_name].sel(time=target_time).values
            
            # --- Get CRA5 Predictions ---
            if c_name in vname_to_channels:
                ch_idx = vname_to_channels[c_name]
                recon_data = big_numpy[ch_idx, :, :]
            else:
                print(f"[ERROR] Cannot find '{c_name}' in CRA5 channels!")
                continue

            error_data = recon_data - gt_data

            # --- Calculate Metrics ---
            rmse = np.sqrt(np.mean(error_data**2))
            mae = np.mean(np.abs(error_data))
            rel_l2 = np.linalg.norm(error_data) / np.linalg.norm(gt_data)

            results_table.append({
                "Variable": c_name,
                "Unit": unit,
                "RMSE": f"{rmse:.4f}",
                "MAE": f"{mae:.4f}",
                "Rel_L2": f"{rel_l2:.2e}"
            })

        print("| Variable | Unit | RMSE | MAE | Rel L2 Error |")
        print("| :--- | :--- | :--- | :--- | :--- |")
        for row in results_table:
            print(f"| {row['Variable']} | {row['Unit']} | {row['RMSE']} | {row['MAE']} | {row['Rel_L2']} |")

    finally:
        if 'bin_path' in locals() and os.path.exists(bin_path):
            os.remove(bin_path)

if __name__ == "__main__":
    test_custom_9_vars("2020-01-01T00:00:00")
```

Hopefully, this helps clear up the discrepancies!
